File: Other Research/Python ffti Algorithm/v5/test_3D_inter.py
import ffti_v6 as fi
import numpy as np
from time import time

x_max, y_max, z_max = 2 * np.pi, 2 * np.pi, 2 * np.pi
Ni, Nj, Nk = 16, 16, 16

x, y, z = np.linspace(0, x_max, Ni), np.linspace(0, y_max, Nj), np.linspace(0, z_max, Nk)

rho = np.zeros(shape = (Ni, Nj, Nk))

def func(x, y, z):
   rho = np.e ** (np.sin(x) + np.sin(y) + np.sin(z))
   return rho

for i in range(0, Ni): 
   for j in range(0, Nj): 
      for k in range(0, Nk):
         rho[i, j, k] = func(x[i], y[j], z[k])
         # print("rho[{0:.5f}][{1:.5f}][{2:.5f}] = {3}".format(x[i], y[j], z[k], rho[i, j, k]))

Ni_inter, Nj_inter, Nk_inter = 4 * Ni, 4 * Nj, 4 * Nk
x_inter, y_inter, z_inter = np.linspace(0, x_max, Ni_inter), np.linspace(0, y_max, Nj_inter), np.linspace(0, z_max, Nk_inter)

a = fi.inter_3D(x, y, z, rho, [1.57, 1.99, 2.52])
b = func(1.57, 1.99, 2.52)
print("rho_int({0:.5f}, {1:.5f}, {2:.5f}) = {3}".format(1.57, 1.99, 2.52, a))
print("rho_ana({0:.5f}, {1:.5f}, {2:.5f}) = {3}".format(1.57, 1.99, 2.52, b))

"""
rho_inter = np.zeros(shape = (Ni_inter, Nj_inter, Nk_inter))

for i in range(0, Ni_inter): 
   for j in range(0, Nj_inter): 
      for k in range(0, Nk_inter):
         rho_inter[i, j, k] = fi.inter_3D(x, y, z, rho, [x_inter[i], y_inter[j], z_inter[k]])
         print("rho_inter[{0}][{1}][{2}] = {3}".format(i, j, k, rho_inter[i, j, k]))

rho_inter_error = np.zeros(shape = (Ni_inter, Nj_inter, Nk_inter))

for i in range(0, Ni_inter): 
   for j in range(0, Nj_inter): 
      for k in range(0, Nk_inter):
         rho_inter_error[i, j, k] = rho_inter[i, j, k] - func(x_inter[i], y_inter[j], z_inter[k])
         print("rho_inter_error[{0}][{1}][{2}] = {3}".format(i, j, k, rho_inter_error[i, j, k]))
"""

File: Other Research/Dead End Projects/C ffti Development/Abandoned/test_3D_inter.c
#include "trifourier_v0.c"
#include <math.h>

double func(double x, double y, double z) {
   return pow(M_E, sin(x) + sin(y) + sin(z));
}

int main() {
   const int Ni = 32, Nj = 32, Nk = 32;
   const double x_max = 2 * M_PI, y_max = 2 * M_PI, z_max = 2 * M_PI;

   double *x = linspace(0, x_max, Ni);
   double *y = linspace(0, y_max, Nj);
   double *z = linspace(0, z_max, Nk);

   double (*rho)[Nj][Nk] = malloc(Ni * Nj * Nk * sizeof(double));

   for (int i = 0; i < Ni; i++) {
      for (int j = 0; j < Nj; j++) {
         for (int k = 0; k < Nk; k++) {
            // rho[i][j][k] = pow(M_E, sin(x[i]) + sin(y[j]) + sin(z[k]));
            rho[i][j][k] = func(x[i], y[j], z[k]);
            // printf("rho[%.5f][%.5f][%.5f] = %.16f\n", x[i], y[j], z[k], rho[i][j][k]);
         }
      }
   }
   
   const int Ni_inter = 4 * Ni, Nj_inter = 4 * Nj, Nk_inter = 4 * Nk;
   double *x_inter = linspace(0, x_max, Ni_inter);
   double *y_inter = linspace(0, y_max, Nj_inter);
   double *z_inter = linspace(0, z_max, Nk_inter);

   double a = inter_3D(x, y, z, Ni, Nj, Nk, rho, 1.57, 1.99, 2.52);
   double b = func(1.57, 1.99, 2.52);
   printf("rho_int(%.5f, %.5f, %.5f) = %.16f\n", 1.57, 1.99, 2.52, a);
   printf("rho_ana(%.5f, %.5f, %.5f) = %.16f\n", 1.57, 1.99, 2.52, b);

/*
   double rho_inter[Ni_inter][Nj_inter][Nk_inter];

   for (int i = 0; i < Ni_inter; i++) {
      for (int j = 0; j < Nj_inter; j++) {
         for (int k = 0; k < Nk_inter; k++) {
            rho_inter[i][j][k] = inter_3D(x, y, z, Ni, Nj, Nk, rho, x_inter[i], y_inter[j], z_inter[k]);
            printf("rho_inter[%d][%d][%d] = %.16f\n", i, j, k, rho_inter[i][j][k]);
         }
      }
   }

   double rho_inter_error[Ni_inter][Nj_inter][Nk_inter];

   for (int i = 0; i < Ni_inter; i++) {
      for (int j = 0; j < Nj_inter; j++) {
         for (int k = 0; k < Nk_inter; k++) {
            rho_inter_error[i][j][k] = rho_inter[i][j][k] - func(x_inter[i], y_inter[j], z_inter[k]);
            printf("rho_inter_error[%d][%d][%d] = %.16f\n", i, j, k, rho_inter_error[i][j][k]);
         }
      }
   }
   */

}